src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -1060,22 +1252,96 @@ def migrate_database(self) -> None:
         # 1. the database is completely empty (not initialized)
-        # 2. the database is not empty, but has never been migrated with alembic
+        # 2. the database is not empty and has been migrated with alembic before
+        # 3. the database is not empty, but has never been migrated with alembic
         #   before (i.e. was created with SQLModel back when alembic wasn't
-        #   used)
-        # 3. the database is not empty and has been migrated with alembic before
-        revisions = self.alembic.current_revisions()
-        if len(revisions) >= 1:
-            if len(revisions) > 1:
+        #   used). We don't support this direct upgrade case anymore.
+        current_revisions = self.alembic.current_revisions()
+        head_revisions = self.alembic.head_revisions()
+        if len(current_revisions) >= 1:
+            # Case 2: the database has been migrated with alembic before. Just
+            # upgrade to the latest revision.
+            if len(current_revisions) > 1:
                 logger.warning(
                     "The ZenML database has more than one migration head "
                     "revision. This is not expected and might indicate a "
                     "database migration problem. Please raise an issue on "
                     "GitHub if you encounter this."
                 )
-            # Case 3: the database has been migrated with alembic before. Just
-            # upgrade to the latest revision.
-            self.alembic.upgrade()
+
+            logger.debug("Current revisions: %s", current_revisions)
+            logger.debug("Head revisions: %s", head_revisions)
+
+            # If the current revision and head revision don't match, a database
+            # migration that changes the database structure or contents may
+            # actually be performed, in which case we enable the backup
+            # functionality. We only enable the backup functionality if the
+            # database will actually be changed, to avoid the overhead for
+            # unnecessary backups.
+            backup_enabled = (
+                self.config.backup_strategy != DatabaseBackupStrategy.DISABLED
+                and current_revisions[0] != head_revisions[0]
+            )
+            backup_location: Optional[Any] = None
+            backup_location_msg: Optional[str] = None
+
+            if backup_enabled:
+                try:
+                    logger.info("Backing up the database before migration.")
+                    (
+                        backup_location_msg,
+                        backup_location,
+                    ) = self.backup_database()
+                except Exception:
+                    logger.exception(
+                        "Failed to backup the database. Please check the logs "
+                        "for more details. The upgrade will continue as usual "
+                        "but no recovery is possible if something goes wrong."
```

Comment:
Hmm, so even if I configure a backup strategy, I might have misconfigured something, and the migration will still continue and possibly fail, without any recourse? Can't we make it fail here, and give users a overrride strategy?

Reply:
done